[dtensor] add op support for nll_loss_forward

tianyu-l opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #118950
* __->__ #118917

This is part of the work to support cross entropy in dtensor.

This PR doesn't support nll_loss computation with input sharded on the channel dimension yet. In that case, redistribution to Replicate is needed in sharding propagation.


cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @wconstab @yf225

tianyu-l committed:
[dtensor] add op support for nll_loss_forward

[ghstack-poisoned]

XilunWu commented:
we can remove `nll_loss_forward_rules` and `nll_loss_backward_rules` in `torch/distributed/_tensor/ops/experimental_ops.py` once this PR is done.

XilunWu commented on test/distributed/_tensor/test_math_ops.py:
```diff
@@ -126,6 +126,43 @@ def test_softmax_with_bwd(self):
+                    dist_x, dist_target, reduction=reduction
+                )
+                if shard_dim == channel_dim:
+                    self.assertTrue(dist_y.placements[0].is_replicate())
```
add `comm_debug` test?

tianyu-l replied:
currently `comm_debug` doesn't count communications happening in sharding prop, probably because the __torch_dispatch__ at op level overrides that at the `comm_debug` level; left a TODO item to fix it

XilunWu commented on test/distributed/_tensor/test_math_ops.py:
```diff
@@ -126,6 +126,43 @@ def test_softmax_with_bwd(self):
+                    self.assertTrue(dist_y.placements[0].is_replicate())
+                    self.assertEqual(dist_y.to_local(), y)
+                else:
+                    if reduction == "none":
+                        output_shard_dim = (
+                            shard_dim if shard_dim < channel_dim else shard_dim - 1
+                        )
+                        self.assertTrue(
+                            dist_y.placements[0].is_shard(dim=output_shard_dim)
+                        )
+                    else:
+                        self.assertTrue(dist_y.placements[0].is_partial())
+                    self.assertEqual(dist_y.full_tensor(), y)
```
add `comm_debug` check

XilunWu commented on torch/distributed/_tensor/ops/math_ops.py:
```diff
@@ -280,6 +282,111 @@ def softmax_bwd_rule(op_schema: OpSchema) -> OutputSharding:
+)
+def nll_loss_forward_strategy(mesh: DeviceMesh, op_schema: OpSchema) -> OpStrategy:
+    assert len(op_schema.args_schema) == 5
+    (
+        input_strategy,
+        target_strategy,
+        weight_strategy,
+        reduction,
+        _,
+    ) = op_schema.args_schema
```
question: what is the argument list of this operator?
Is this the associated operator? https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/native_functions.yaml#L11374-L11377

If so, why is the code (https://github.com/pytorch/pytorch/blob/main/torch/_refs/nn/functional/__init__.py#L652-L658) has argument `reduction` of type `str`?
cc @wanchaol

tianyu-l replied:
iiuc the computation happens in C++ at
https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/LossNLL.cpp#L678

and the code pointers should be 
https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/native_functions.yaml#L11392
https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/native_functions.yaml#L11423

not sure when the PyTorch decompositions are called...

wanchaol replied:
nn.functional have potentially different signatures with the aten op, what we should focus on is the aten op instead as that's where we overwrite the ops, the nn.functional would under the hood decompose to a set of aten ops if necessary

However we should test against nn.functional as that's the most user is using

tianyu-l committed:
Update on "[dtensor] add op support for nll_loss_forward"


This is part of the work to support cross entropy in dtensor.

This PR doesn't support nll_loss computation with input sharded on the channel dimension yet. In that case, redistribution to Replicate is needed in sharding propagation.


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 wconstab yf225

[ghstack-poisoned]

XilunWu commented on torch/distributed/_tensor/ops/utils.py:
```diff
@@ -118,6 +118,22 @@ def is_tensor_shardable(shape: Sequence[int], spec: DTensorSpec) -> bool:
+            shards_map[shard_dim] *= spec.mesh.size(i)
+
+    for i, dim_size in enumerate(shape):
+        if shards_map[i] > 1 and dim_size % shards_map[i]:
+            return False
```
This is correct in C/C++ but not in Python:

```
>>> 5 % 5 == True
False
>>> 5 % 2 == True
True
>>> 5 % 3 == True
False
```

should be `(dim_size % shards_map[i] != 0)`

XilunWu reviewed (changes requested):
Overall LGTM! 2 suggestions, 1 question, and 1 change request.

tianyu-l commented:
The failure says "RuntimeError: Cannot use ReduceOp.AVG with Gloo". Not sure how to deal with it...

tianyu-l committed:
Update on "[dtensor] add op support for nll_loss_forward"


This is part of the work to support cross entropy in dtensor.

This PR doesn't support nll_loss computation with input sharded on the channel dimension yet. In that case, redistribution to Replicate is needed in sharding propagation.


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 wconstab yf225

[ghstack-poisoned]

tianyu-l committed:
Update on "[dtensor] add op support for nll_loss_forward"


This is part of the work to support cross entropy in dtensor.

This PR doesn't support nll_loss computation with input sharded on the channel dimension yet. In that case, redistribution to Replicate is needed in sharding propagation.


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 wconstab yf225

[ghstack-poisoned]

wanchaol commented:
> The failure says "RuntimeError: Cannot use ReduceOp.AVG with Gloo". Not sure how to deal with it...

This is because gloo does not support avg reduction, I think we should only run the tests with NCCL/CUDA

XilunWu commented:
As Wanchao said, you can add a check in the for-loop:
if the device is CPU and the reduction is "mean", then "continue".

wanchaol commented on torch/distributed/_tensor/ops/math_ops.py:
```diff
@@ -280,6 +282,110 @@ def softmax_bwd_rule(op_schema: OpSchema) -> OutputSharding:
+        # target doesn't have channel dim, and it follows input on other dims
+        target_placement_strategy = target_strategy.strategies[idx]
+        target_src_spec = target_placement_strategy.output_spec
+        target_tgt_spec = DTensorSpec(
```
nit: might be call it as `target_expected_spec`? this name looks confusing haha

wanchaol commented on torch/distributed/_tensor/ops/math_ops.py:
```diff
@@ -280,6 +282,110 @@ def softmax_bwd_rule(op_schema: OpSchema) -> OutputSharding:
+
+        # make sure input is replicated along the channel dim
+        input_src_spec = input_placement_strategy.output_spec
+        input_tgt_spec = DTensorSpec(
```
ditto: `input_expected_spec`

wanchaol commented on torch/distributed/_tensor/ops/math_ops.py:
```diff
@@ -280,6 +282,110 @@ def softmax_bwd_rule(op_schema: OpSchema) -> OutputSharding:
+        if weight_strategy is not None:
+            assert isinstance(weight_strategy, OpStrategy)
+            weight_src_spec = weight_strategy.strategies[idx].output_spec
+            weight_tgt_spec = DTensorSpec(
```
`weight_expected_spec`

wanchaol commented on torch/distributed/_tensor/ops/math_ops.py:
```diff
@@ -280,6 +282,110 @@ def softmax_bwd_rule(op_schema: OpSchema) -> OutputSharding:
+                    target_tgt_spec.shape, target_tgt_spec
+                ):
+                    raise ValueError(
+                        "The intermediate results of nll_loss cannot be evenly sharded, \
```
> "The intermediate results" 

what intermediate results referring to here?

tianyu-l replied:
It's referring to the results with `reduction='none'`. When `reduction='mean'`, there is an extra step to collapse the big tensor into a scalar.

wanchaol reviewed (approved):
lgtm, have some comments inlined

tianyu-l committed:
Update on "[dtensor] add op support for nll_loss_forward"


This is part of the work to support cross entropy in dtensor.

This PR doesn't support nll_loss computation with input sharded on the channel dimension yet. In that case, redistribution to Replicate is needed in sharding propagation.


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 wconstab yf225

[ghstack-poisoned]

tianyu-l committed:
Update on "[dtensor] add op support for nll_loss_forward"


This is part of the work to support cross entropy in dtensor.

This PR doesn't support nll_loss computation with input sharded on the channel dimension yet. In that case, redistribution to Replicate is needed in sharding propagation.


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 wconstab yf225

[ghstack-poisoned]

tianyu-l committed:
Update on "[dtensor] add op support for nll_loss_forward"


This is part of the work to support cross entropy in dtensor.

This PR doesn't support nll_loss computation with input sharded on the channel dimension yet. In that case, redistribution to Replicate is needed in sharding propagation.


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 wconstab yf225

[ghstack-poisoned]

tianyu-l committed:
Update on "[dtensor] add op support for nll_loss_forward"


This is part of the work to support cross entropy in dtensor.

This PR doesn't support nll_loss computation with input sharded on the channel dimension yet. In that case, redistribution to Replicate is needed in sharding propagation.


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 wconstab yf225

[ghstack-poisoned]

tianyu-l commented:
@pytorchbot merge
